File: MP_2017_lato/lista_12/zad5.hs
data Cyclist a= Elem (Cyclist a) a (Cyclist a) deriving (Show)


fromList:: [a] -> Cyclist a

fromList (x:xs) = res
  where
      res = Elem last x next
      (next, last) = aux res xs res

aux prev [] first = (first, prev)
aux prev (x:xs) first = (this, last)
  where
      this = Elem prev x next
      (next, last) = aux this xs first


forward:: Cyclist a -> Cyclist a
forward ( Elem _ _ x) =x


backward:: Cyclist a -> Cyclist a
backward (Elem x _ _) = x

label:: Cyclist a -> a
label (Elem _ x _) =x


enumInts :: Cyclist Integer

enumInts = Elem neg 0 pos where
    pos= positive enumInts 1
    neg = negative enumInts (-1)

positive :: (Num a) => Cyclist a -> a -> Cyclist a
positive prev curr = this where
    this= Elem prev curr (positive this (curr+1))

negative :: (Num a) => Cyclist a -> a -> Cyclist a
negative next curr = this where
    this = Elem (negative this (curr-1)) curr next

File: MP_2017_lato/lista_3/zad5.pl
insert([], Elem, [Elem]) :- !.
insert([H|T], Elem, Res) :- Elem<H,
                            !,
                            Res=[Elem, H|T].
insert([H|T], Elem, [H|S]) :- insert(T, Elem, S).

ins_sort([X],[X]) :- !.
ins_sort([H|T], Res) :- ins_sort(T, Res1),
                        insert(Res1,H,Res).
